test/contrib/kubernetes_test.py

```diff
@@ -86,15 +85,16 @@ def test_success_job(self):
     def test_fail_job(self):
         fail = FailJob()
         self.assertRaises(RuntimeError, fail.run)
-        # Check for retrials
-        kube_api = HTTPClient(KubeConfig.from_file("~/.kube/config"))  # assumes minikube
-        jobs = Job.objects(kube_api).filter(selector="luigi_task_id="
-                                                     + fail.job_uuid)
-        self.assertEqual(len(jobs.response["items"]), 1)
-        job = Job(kube_api, jobs.response["items"][0])
-        self.assertTrue("failed" in job.obj["status"])
-        self.assertTrue(job.obj["status"]["failed"] > fail.max_retrials)
-        self.assertTrue(job.obj['spec']['template']['metadata']['labels'] == fail.labels())
+        # TODO: Discuss/consider what to re-implement here, possibly only keep the fail.labels check, possibly ad backoff limit and name check
```

Comment:
in terms of what to test here, i'd suggest ensuring that a failing job behaves as expected. If that includes attempting a retry, go for it, etc.

Reply:
**TL;DR**: Luigi isn't "managing Kubernetes" so the above tests don't make sense to test.  We're telling Kubernetes contractually what to do (eg: number of retries) and it tries to do that.  No purpose in testing what Kubernetes actually did, the output of what it actually do can be wildly different dependant on many variables and situational aspects.  Kubernetes manages the aspect of retrials, which I think is a large contrast to how Luigi works with many other providers manually re-submitting/executing/scheduling things.

Well... so @dlstadther what you're asking is for me to write tests to ensure Kubernetes is executing what you tell it to. 
 I'm trying to think of a good metaphor... but in a sense, it's like someone going to a master at their craft, lets say a baker in this example, and going into their kitchen and setting up some timers and some cameras to inspect the process at every step of the way to ensure the inputs end up with the same outputs.  I think the way to explain this as well, is Luigi isn't "managing Kubernetes" because there's no such thing really (well, sure, controllers do this, but we aren't a controller for Kubernetes).  I stripped this silly retry code that was in this contrib that attempted to fiddle with some retry logic because all of that is 100% built-into Kubernetes and makes no sense.  And thus, because we are not shepherding things in Kubernetes, we aren't "managing" kubernetes any more, we also shouldn't be testing things which are managed by Kubernetes.
 
 By defining the yaml object properly and by Kubernetes accepting that object, it has committed to _trying_ to execute that contract to the best of its ability, within the constraints you give it.  And contractually it tries to do that.  If you tell it to retry x times, it will do everything it can to ensure that happens.  However, there are various various reasons (eg: resource limits/constraints, node selectors, taints, etc) that it can't fulfill the request contractually.  Since we're defining the yaml to include a specific number of retries, kubernetes will _attempt_ to do that.  But checking if Kubernetes _did_ retry, doesn't make a lot of sense.  It either did or it didn't, that's not our responsibility to test against.  In theory, those above tests are actually unstable depending on the stability of the cluster, nodes in the cluster, resource constraints (or lack of), oomkiller happening, unable to scale up the cluster, etc.  A job could absolutely be failed, and not retry at all because it was never able to initialize because of a mismatch of nodelabels, or taints, etc.
 
 I think what might make sense here is to test that everything we tried to set, is actually set.  Eg: number of max retrials, image of the job, job definition details, etc.  That's the best thing I can think of, and I hope my explanation makes sense above.

Reply:
But that's why I left the comment that said this... input needed here.   :)  Shall I just add some tests to check if what we set is actually set in the returned yaml object?  That seems... pedantic, but might be valuable at some point in the future if the API changes in Kubernetes that might make our expectations change (eg: renaming or moving where we set a variable, such as the retries)

Reply:
@AndrewFarley , i appreciate your thoroughness of response. Yes, i'm familiar with Kubernetes (less so with how we're using Luigi here to interact; although looks moderately straight forwards as a k8s job that we then just poll for its status). So when i say we should "ensure that a failing job behaves as expected", I'm really referring to ensuring that Luigi acknowledges a failing job and responds to it appropriately. What I don't want to see is that a Luigi job thinks it was successful when its respective Kubernetes job actually exited in an erroneous state (in state phase error, or possibly even evicted by the control plane and thus an unknown status).

In the case that a job is not found, it looks like `__get_job` will raise an AssertionError; whereas if the job state is "Failed" (from `__get_job_status`), then we'll still raise an AssertionError within `_track_job`).

So the place I think we're missing test coverage for (that we should) is to ensure we get an AssertionError when its job fails. That isn't testing K8s functions correctly; that's ensuring that our Luigi code identifies and responds to the k8s api responses correctly and doesn't falsely think we were successful or respond to error in a way we don't expect. 

The extent to which we do this can be a little checking that:
```
        fail = FailJob()
        self.assertRaises(RuntimeError, fail.run)
```
and leaving it at that. 

So effectively removing the commented out section and calling it a day.

Reply:
Resolved that above issue, simple failjob